lib/ecto/schema.ex

```diff
@@ -1405,10 +1412,44 @@ defmodule Ecto.Schema do
   defp type_to_module(:utc_datetime), do: DateTime
   defp type_to_module(other), do: other
 
+  defp replace_field_sources(fields = %{}, aliases) do
+    Enum.reduce(fields, %{}, fn
+      {field, value}, acc ->
+        field = case is_atom(field) do
+                  true -> field
+                  false -> String.to_atom(field)
+                end
+
+        field_name = case List.keyfind(aliases, field, 1) do
+                       {aliased, _} -> to_string(aliased)
+                       nil -> field
+                     end
+
+        Map.put(acc, field_name, value)
+    end)
+  end
+  defp replace_field_sources({fields, values}, aliases) do
+    {replace_field_sources(fields, [], aliases), values}
+  end
+  defp replace_field_sources([field|fields], list, aliases) do
+    field = case List.keyfind(aliases, field, 1) do
+              {field, _} -> field
+              _ -> field
+            end
+
+    replace_field_sources(fields, list ++ [field], aliases)
+  end
+  defp replace_field_sources([], list, _aliases) do
+    list
+  end
+
   @doc false
   def __load__(schema, prefix, source, context, data, loader) do
     struct = schema.__struct__()
     types = schema.__schema__(:types)
+    aliases = schema.__schema__(:aliases)
+
+    data = replace_field_sources(data, aliases)
```

Comment:
This adds another iteration over the data on every load. It will have a significant impact on performance. Is there maybe a way to do this in one pass?